[InstCombine] lshr (mul (X, 2^N + 1)), N -> X when X is half-width

AZero13 opened the pull request:
Alive2 Proof:
https://alive2.llvm.org/ce/z/Yd2CKF

nikic commented on llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:
```diff
@@ -1463,11 +1463,17 @@ Instruction *InstCombinerImpl::visitLShr(BinaryOperator &I) {
+        KnownBits KnownX =
+            computeKnownBits(X, /* Depth */ 0, cast<Instruction>(Op0));
+        if (KnownX.countMaxActiveBits() <= ShAmtC)
+          return replaceInstUsesWith(I, X);
```
I don't understand the purpose of this special case. If the top bits are all zero, then the generated `lshr(X,N)` will fold to zero and the result will be just `X`.

I think the only additional fold this would enable is the extra-use case, which is not a case we're interested int.

nikic commented:
If you want to create a draft pull request, please **create it as a draft**. Do not first open it as a normal pull request and then convert it into a draft. Otherwise you will still notify and subscribe everyone.

AtariDreams commented:
I have used this platform for years and today I learned you could do that instead of converting to draft. I'm so sorry I feel so dumb.

topperc commented on llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:
```diff
@@ -1465,9 +1465,9 @@ Instruction *InstCombinerImpl::visitLShr(BinaryOperator &I) {
           MulC->logBase2() == ShAmtC) {
         // Look for a "splat" mul pattern - it replicates bits across each half
         // of a value, so a right shift is just a mask of the low bits:
```
Comment needs to be updated. "mask of the low bits" describes the AND that was inserted before.

AZero13 committed:
[InstCombine] lshr (mul (X, 2^N + 1)), N -> X when X is half-width

Alive2 Proofs:
https://alive2.llvm.org/ce/z/eSinJY
https://alive2.llvm.org/ce/z/sweDgc
https://alive2.llvm.org/ce/z/-2dXZi
https://alive2.llvm.org/ce/z/kx2PhF
https://alive2.llvm.org/ce/z/e9QjM6

nikic reviewed (approved):
LGTM
